lang/csharp/src/apache/main/Util/Decimal.cs

```diff
@@ -0,0 +1,137 @@
+
+            var precisionVal = schema.GetProperty("precision");
+
+            if (string.IsNullOrEmpty(precisionVal))
+                throw new AvroTypeException("'decimal' requires a 'precision' property");
+
+            var precision = int.Parse(precisionVal, CultureInfo.CurrentCulture);
+
+            if (precision <= 0)
+                throw new AvroTypeException("'decimal' requires a 'precision' property that is greater than zero");
+
+            var scale = GetIntPropertyValueFromSchema(schema, "scale");
+
+            if (scale < 0 || scale > precision)
+                throw new AvroTypeException("'decimal' requires a 'scale' property that is zero or less than or equal to 'precision'");
+        }
+
+        /// <inheritdoc/>      
+        public override object ConvertToBaseValue(object logicalValue, LogicalSchema schema)
+        {
+            var decimalValue = (AvroDecimal)logicalValue;
+            var logicalScale = GetIntPropertyValueFromSchema(schema, "scale");
+            var scale = decimalValue.Scale;
+
+            if (scale != logicalScale)
+                throw new ArgumentOutOfRangeException(nameof(logicalValue), $"The decimal value has a scale of {scale} which cannot be encoded against a logical 'decimal' with a scale of {logicalScale}");
+
+            var buffer = decimalValue.UnscaledValue.ToByteArray();
+
+            if (BitConverter.IsLittleEndian)
+            {
+                Array.Reverse(buffer);
+            }
+
+            return Schema.Type.Bytes == schema.BaseSchema.Tag
+                ? (object)buffer
+                : (object)new GenericFixed(
+                    (FixedSchema)schema.BaseSchema,
+                    GetDecimalFixedByteArray(buffer, ((FixedSchema)schema.BaseSchema).Size,
+                    decimalValue.Sign < 0 ? (byte)0xFF : (byte)0x00));
+        }
+
+        /// <inheritdoc/>
+        public override object ConvertToLogicalValue(object baseValue, LogicalSchema schema)
+        {
+            var buffer = Schema.Type.Bytes == schema.BaseSchema.Tag
+                ? (byte[])baseValue
+                : ((GenericFixed)baseValue).Value;
+
+            if (BitConverter.IsLittleEndian)
+            {
+                Array.Reverse(buffer);
+            }
+
+            return new AvroDecimal(new BigInteger(buffer), GetIntPropertyValueFromSchema(schema, "scale"));
+        }
+
+        /// <inheritdoc/>
+        public override Type GetCSharpType(bool nullible)
+        {
+            return nullible ? typeof(AvroDecimal?) : typeof(AvroDecimal);
+        }
+
+        /// <inheritdoc/>
+        public override bool IsInstanceOfLogicalType(object logicalValue)
+        {
+            return logicalValue is AvroDecimal;
+        }
+
+        private static int GetIntPropertyValueFromSchema(Schema schema, string propertyName, int defaultVal = 0)
+        {
+            var scaleVal = schema.GetProperty(propertyName);
```

Comment:
`scaleVal` is an inappropriate name here. i guess the result of a refactoring.

... also, the method is private, and only used for the scale property so i would just make this method more specific: `GetScalePropertyValueFromSchema` and only have the `Schema` arg). If you do keep it more general, I think having a default for `defaultVal` acts to obfuscate rather than help, so i'd remove that. But I do think the scale specific method is most appropriate.